File: HW4/HW4Tests/CalculatorTests.fs
module HW4Tests.CalculatorTests

open Xunit
open HW4

[<Theory>]
[<InlineData(6, 9, 15)>]
[<InlineData(17, -7, 10)>]
[<InlineData(-2, 9, 7)>]
[<InlineData(-3, -5, -8)>]
let ``Calculate_Plus_WillReturnCorrectResult`` (val1, val2, expected) =
    let result = Calculator.calculate Operations.Plus val1 val2
    Assert.Equal(expected, result)
    
[<Theory>]
[<InlineData(5, 7, -2)>]
[<InlineData(10, -7, 17)>]
[<InlineData(-7, 7, -14)>]
[<InlineData(-25, -36, 11)>]
let ``Calculate_Minus_WillReturnCorrectResult`` (val1, val2, expected) = 
    let result = Calculator.calculate Operations.Minus val1 val2
    Assert.Equal(expected, result)
    
[<Theory>]
[<InlineData(4, 3, 12)>]
[<InlineData(6, -3, -18)>]
[<InlineData(-5, 6, -30)>]
[<InlineData(-7, -7, 49)>]
let ``Calculate_Multiply_WillReturnCorrectResult`` (val1, val2, expected) = 
    let result = Calculator.calculate Operations.Multiply val1 val2
    Assert.Equal(expected, result)

[<Theory>]
[<InlineData(6, 2, 3)>]
[<InlineData(7, -1, -7)>]
[<InlineData(-10, 5, -2)>]
[<InlineData(-24, -6, 4)>]
[<InlineData(7, 2, 3.5)>]
let ``Calculate_Divide_WillReturnCorrectResult`` (val1,  val2, expected) = 
    let result = Calculator.calculate Operations.Divide val1 val2
    Assert.Equal(expected, result)

File: HW8/HW8Tests/CalculatorTests.cs
using HW8.Controllers;
using HW8.Logic;
using Xunit;

namespace HW8Tests
{
    public class CalculatorTests
    {
        [Theory]
        [InlineData("4", "plus", "3", "7")]
        [InlineData("32", "minus", "2", "30")]
        [InlineData("6", "division", "5", "1.2")]
        [InlineData("4,5", "multiplication", "2", "9")]
        public void Calculate_ValidArguments(string v1, string op, string v2, string excepted)
        {
            MakeGeneralPartTests(v1, op, v2, excepted);
        }
        
        [Theory]
        [InlineData("error1", "plus", "7", "Value is not double: error1")]
        [InlineData("44", "minus", "error2", "Value is not double: error2")]
        [InlineData("error1", "minus", "error2", "Value is not double: error1 and Value is not double: error2")]
        [InlineData("64", "error", "88", "The calculator does not recognize this operation: error")]
        [InlineData("66", "division", "0", "You can't divide by zero")]
        [InlineData("0", "division", "0", "The result is undefined")]
        public void Calculate_InvalidArguments(string v1, string op, string v2, string excepted)
        {
            MakeGeneralPartTests(v1, op, v2, excepted);
        }
        
        private static void MakeGeneralPartTests(string v1, string op, string v2, string excepted)
        {
            var result = new CalculatorController()
                .Calculate(new Calculator(), v1, op, v2);
            Assert.Equal(excepted, result);
        }
    }
}